Row and Column Iterator for Matrices

arnavs opened the pull request:
Talked in Slack about creating `eachrow(m::Matrix)` and `eachcolumn(m::Matrix)` which return an iterator over the rows and columns of `M`. Here's a stab at it.

Commit:
first pass, sans tests

Commit:
add test

Commit:
refactor a bit

Commit:
fix eltype to show views

arnavs commented:
(The whitespace check is failing; will fix.)

Commit:
fix whitespace

Commit:
fix typo

ararslan commented on stdlib/LinearAlgebra/src/LinearAlgebra.jl:
```diff
@@ -67,6 +67,7 @@ export
     copyto!,
     copy_transpose!,
     cross,
+    cols,
```
I'd call it `columns`. We typically spell things out rather than abbreviating when they aren't too onerous to type.

fredrikekre commented:
See #14491

arnavs commented:
Thanks @fredrikekre, I updated the naming convention to match what was happening in that thread (`eachrow` and `eachcolumn`).  

Also, what's going on with CI? Looks like we have two errors in something called `sockets` for 2 of the builds?

Commit:
rename [ci skip]

Commit:
typofix [ci skip]

Commit:
add eachslice() [ci skip]

Commit:
eachcolumn => eachcol and a bit of error handling [ci skip]

andyferris commented:
Cool, nice, I was hoping we would implement something like this soon!

This one is a little close to my heart. For this functionality, I was hoping we could be slightly more generic - by being able to slice up arrays of any dimensionality by any one (or more) dimension.

Additionally, since arrays naturally support random-access, it's also possible to turn the "iterator"s into a fully-fledged `AbstractArray`s in their own right. That is, `eachrow` of a matrix could return a nested vector of vectors. This would let users interact with the result in ways beyond a simple `for` loop. For example, the presence of the indices makes it easier to match up multiple containers which share indices along certain dimensions. Finally, different array types may have their own implementations of `map`, `reduce` and so-on which is quite different to a sequential `for` loop over all elements (consider for example doing something in parallel with the rows of a `DMatrix` from DistributedArrays).

Anyway, because I've been interested in this for a little while, I implemented such a nested array dimension splitting in a package called [SplitApplyCombine](https://github.com/JuliaData/SplitApplyCombine.jl) under the functions `splitdims` and `splitdimsview` (the latter being lazy). My intention here was always to develop these in a package and hopefully move these to `Base` through a PR here for Julia 1.x. (It should be noted that I don't *particularly* like my function names, but they are at least descriptive placeholders for now).

If I may be slightly philosophical - in my view, it seems more productive and flexible to let users put together different "split", "apply" and "combine" strategies like lego bricks, rather than to rely on increasingly complex functions like `mapslices` to do a split-apply-combine strategy all in one step, which is why I'm advocating for this in general, and excited by this PR in particular.

Anyway, if @arnavs and the community are at all interested, I'd be happy to help expand the functionality here to work on arbitrary dimensions and return a fully-fledged `AbstractArray`.

andyferris commented:
Cool, nice, I was hoping we would implement something like this soon!

This one is a little close to my heart. For this functionality, I was hoping we could be slightly more generic - by being able to slice up arrays of any dimensionality by any one (or more) dimension.

Additionally, since arrays naturally support random-access, it's also possible to turn the "iterator"s into a fully-fledged `AbstractArray`s in their own right. That is, `eachrow` of a matrix could return a nested vector of vectors. This would let users interact with the result in ways beyond a simple `for` loop. For example, the presence of the indices makes it easier to match up multiple containers which share indices along certain dimensions. Finally, different array types may have their own implementations of `map`, `reduce` and so-on which is quite different to a sequential `for` loop over all elements (consider for example doing something in parallel with the rows of a `DMatrix` from DistributedArrays).

Anyway, because I've been interested in this for a little while, I implemented such a nested array dimension splitting in a package called [SplitApplyCombine](https://github.com/JuliaData/SplitApplyCombine.jl) under the functions `splitdims` and `splitdimsview` (the latter being lazy). My intention here was always to develop these in a package and hopefully move these to `Base` through a PR here for Julia 1.x. (It should be noted that I don't *particularly* like my function names, but they are at least descriptive placeholders for now).

If I may be slightly philosophical - in my view, it seems more productive and flexible to let users put together different "split", "apply" and "combine" strategies like lego bricks, rather than to rely on increasingly complex functions like `mapslices` to do a split-apply-combine strategy all in one step, which is why I'm advocating for this in general, and excited by this PR in particular.

Anyway, if @arnavs and the community are at all interested, I'd be happy to help expand the functionality here to work on arbitrary dimensions and return a fully-fledged `AbstractArray`.

andyferris commented:
Cool, nice, I was hoping we would implement something like this soon!

This one is a little close to my heart. For this functionality, I was hoping we could be slightly more generic - by being able to slice up arrays of any dimensionality by any one (or more) dimension.

Additionally, since arrays naturally support random-access, it's also possible to turn the "iterator"s into a fully-fledged `AbstractArray`s in their own right. That is, `eachrow` of a matrix could return a nested vector of vectors. This would let users interact with the result in ways beyond a simple `for` loop. For example, the presence of the indices makes it easier to match up multiple containers which share indices along certain dimensions. Finally, different array types may have their own implementations of `map`, `reduce` and so-on which is quite different to a sequential `for` loop over all elements (consider for example doing something in parallel with the rows of a `DMatrix` from DistributedArrays).

Anyway, because I've been interested in this for a little while, I implemented such a nested array dimension splitting in a package called [SplitApplyCombine](https://github.com/JuliaData/SplitApplyCombine.jl) under the functions `splitdims` and `splitdimsview` (the latter being lazy). My intention here was always to develop these in a package and hopefully move these to `Base` through a PR here for Julia 1.x. (It should be noted that I don't *particularly* like my function names, but they are at least descriptive placeholders for now).

If I may be slightly philosophical - in my view, it seems more productive and flexible to let users put together different "split", "apply" and "combine" strategies like lego bricks, rather than to rely on increasingly complex functions like `mapslices` to do a split-apply-combine strategy all in one step, which is why I'm advocating for this in general, and excited by this PR in particular.

Anyway, if @arnavs and the community are at all interested, I'd be happy to help expand the functionality here to work on arbitrary dimensions and return a fully-fledged `AbstractArray`.

andyferris commented:
Cool, nice, I was hoping we would implement something like this soon!

This one is a little close to my heart. For this functionality, I was hoping we could be slightly more generic - by being able to slice up arrays of any dimensionality by any one (or more) dimension.

Additionally, since arrays naturally support random-access, it's also possible to turn the "iterator"s into a fully-fledged `AbstractArray`s in their own right. That is, `eachrow` of a matrix could return a nested vector of vectors. This would let users interact with the result in ways beyond a simple `for` loop. For example, the presence of the indices makes it easier to match up multiple containers which share indices along certain dimensions. Finally, different array types may have their own implementations of `map`, `reduce` and so-on which is quite different to a sequential `for` loop over all elements (consider for example doing something in parallel with the rows of a `DMatrix` from DistributedArrays).

Anyway, because I've been interested in this for a little while, I implemented such a nested array dimension splitting in a package called [SplitApplyCombine](https://github.com/JuliaData/SplitApplyCombine.jl) under the functions `splitdims` and `splitdimsview` (the latter being lazy). My intention here was always to develop these in a package and hopefully move these to `Base` through a PR here for Julia 1.x. (It should be noted that I don't *particularly* like my function names, but they are at least descriptive placeholders for now).

If I may be slightly philosophical - in my view, it seems more productive and flexible to let users put together different "split", "apply" and "combine" strategies like lego bricks, rather than to rely on increasingly complex functions like `mapslices` to do a split-apply-combine strategy all in one step, which is why I'm advocating for this in general, and excited by this PR in particular.

Anyway, if @arnavs and the community are at all interested, I'd be happy to help expand the functionality here to work on arbitrary dimensions and return a fully-fledged `AbstractArray`.

andyferris commented:
(Github had problems earlier today... looks like they posted all my retries!)

joshday commented:
I'll add that I've also implemented `eachrow`/`eachcol` in OnlineStatsBase.

The difference from the implementation in this PR is that OnlineStatsBase uses a buffer instead of views (I think views were a little slower, but that could be investigated further).  The code is here: https://github.com/joshday/OnlineStatsBase.jl/blob/0449a586f9229f6bf43a2214b11a271318a240a4/src/OnlineStatsBase.jl#L84-L142.

arnavs commented:
Replies to feedback (first @andyferris, then @joshday): 

1. Fully agree, would love if the slicing API was n-dimensional. 

2. > Additionally, since arrays naturally support random-access, it's also possible to turn the "iterator"s into a fully-fledged AbstractArrays in their own right. That is, eachrow of a matrix could return a nested vector of vectors.
    
    For sure, we could implement `rows(M) = collect(eachrow(M))` or something similar. Is that what you had in mind?

3. > Anyway, because I've been interested in this for a little while, I implemented such a nested array dimension splitting in a package called SplitApplyCombine under the functions splitdims and splitdimsview 

    Sounds good, I'll check this out. 

4. > Anyway, if @arnavs and the community are at all interested, I'd be happy to help expand the functionality here to work on arbitrary dimensions and return a fully-fledged AbstractArray.

      For sure. Maybe if people think this is a good idea (cc: @ararslan, @StefanKarpinski, @fredrikekre), we can merge this and then implement a method `eachslice(A::AbstractArray)`? Or, if we need to workshop this a bit, that's OK too.

5. > I'll add that I've also implemented eachrow/eachcol in OnlineStatsBase...I think views were a little slower...

    Quite possible, but that's above my pay grade. I'll let others weigh in.

StefanKarpinski commented:
@mbauman, can you take a look at this?

mbauman commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,50 @@ function normalize(v::AbstractVector, p::Real = 2)
+    if state > it.length
+        return nothing
+    elseif it.row == true
+        obj = @view it.M[state, :]
```
This is type-unstable.  Both branches return views, yes, but views are specially parameterized by their indices to ensure fast accesses.  Here this branch returns a SubArray with indices `Tuple{Int, Colon}` and in the other branch the indices are `Tuple{Colon, Int}`.  It's especially important for the individual elements returned from the iterator to be inferred properly as if it's unstable then users will have to pay the cost of dynamic dispatch in every iteration of their loop.

You can test this interactively with `@code_warntype` or in the tests file with `@test @inferred(first(eachrow(A))) == …`.

mbauman commented:
I agree with most all of what @andyferris wrote above, except that I'd suggest we start as simple as possible.  I'll also note that there's another package implementation in JuliennedArrays.jl.

I'd suggest simply:

```julia
eachrow(A::AbstractArray) = (view(A, i, :) for i in indices(A, 1))
eachcol(A::AbstractArray) = (view(A, :, j) for j in indices(A, 2))
```

For the general case, it looks like constant propagation can _almost_ handle:

```julia
eachslice(A, d) = (view(A, ntuple(n->n==d ? i : (:), ndims(A))...) for i in axes(A, d))
```

but we're not quite there yet.  It'll take a bit more to make this type-stable.

mbauman commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,50 @@ function normalize(v::AbstractVector, p::Real = 2)
 end
+
+# Iterator for the rows() and cols()
+struct MatrixDimIterator{T}
```
If you parameterize by the parent array type itself, then you can allow any `AbstractMatrix` to be iterated over — and not just `Matrix`.  This includes things like `ReshapedArray`s and other views — things that we try hard to ensure work as much as possible.  `P` is a common variable for this as this can be thought of as the "parent" array that generates the behavior here.

```julia
struct MatrixDimIterator{P<:AbstractMatrix}
…
    M::P
```

mbauman commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,50 @@ function normalize(v::AbstractVector, p::Real = 2)
+Get an iterator for the rows of M.
+See also [`eachcol`](@ref) and [`eachslice`](@ref).
+"""
+eachrow(M::Matrix) = MatrixDimIterator(true, size(M, 1), M)
```
With the above change, these functions could accept all `AbstractMatrix`es.

arnavs replied:
Thanks @mbauman for feedback. So do you think it's worth iterating on this to incorporate the changes? Or just closing the PR and waiting to merge in something from a package

mbauman replied:
I'd suggest just simplifying this to the generator definitions I wrote above (https://github.com/JuliaLang/julia/pull/29749#issuecomment-432015741), but let's see what others think first.

arnavs replied:
Sounds good. Committed the changes, with a clarification that `eachrow` and `eachcol` are only written for `AbstractVecOrMat`.

mbauman reviewed (commented):
I'd really like to see this as a more general structure, but I've added a few pedagogical comments based on the code you've written that I hope will be helpful.

mbauman commented:
> buffer instead of views

I think I prefer the views behavior.  Yes, buffers can be faster in many situations, but they also can have strange effects if you, e.g., `collect` the iterator.  I prefer the view behaviors, and there's hope that someday they'll be faster (#14955).

andyferris commented:
>  we can merge this and then implement a method `eachslice(A::AbstractArray)`

> except that I'd suggest we start as simple as possible

I agree - let's do baby steps! :)

My concern is more on the bikeshedding front - that a function name such as `eachrow` is great for an iterator (`for row in eachrow(matrix)` is indeed very readable), but IMO not so great for a general function which transforms an array into another array. If we know the function will be expanded to included more functionality later, we may as well rename it now to reduce churn.

Our typical policy is to pick a simple English verb for Julia functions (and I don't see why we would deviate from that). Ideas that come to mind include `slice`, `split`, `splitdims`.

> I'll also note that there's another package implementation in JuliennedArrays.jl.

Yes, I meant to come back and acknowledge JuliennedArrays (sorry @bramtayl). While I am talking function names - while I *really* appreciate the `julienne` play on words (genius!), for Julia `Base` perhaps a simpler English verb might be preferrable for new users (especially for those whose mother tongue isn't English)?

Commit:
feedback

Commit:
fix whitespace

andyferris commented:
@arnavs wrote:

> For sure, we could implement rows(M) = collect(eachrow(M)) or something similar. Is that what you had in mind? 

To be clear, I was suggesting that `eachrow` is itself a lazy view of the rows of the matrix, such that `eachrow(::AbstractMatrix{T}) --> AbstractVector{<:AbstractVector{T}}`. (Except perhaps renaming `eachrow` to a verb).

> (Edit: I still think it's valuable to keep the lightweight view iterator for instances that aren't conducive to reproducing all the data.)

I'm not sure if I understood this comment correctly.

The `AbstractArray` wrapper won't be any more "heavyweight" in the sense that will remain a very simple `struct` wrapper (like `SubArray` or `Generator`). It will however support the `AbstractArray` interface, making it far more useful than an iterator or a `Generator`. There needn't be a full copy made, or anything like that, if that is what you mean by "reproduce". (E.g. I prefer the functionality of  `SplitApplyCombine.splitdimsview` over `SplitApplyCombine.splitdims`. The latter can be achieved by a `copy` of the former).

bramtayl commented:
I'm always surprised by how much of a kick people get out of the name julienne. But yeah I'm a fan of small words too. Can we haz programming language in esperanto?

arnavs commented:
Hey @mbauman just wanted to put this back on your radar.

mbauman commented:
There are two things on which I'd like to gather a bit more of a consensus:

1. Naming: personally I like `eachslice` / `eachrow`/ `eachcol`. I like the meaningful shorthands. I actually like the limited behaviors they imply. An alternative would be `slices` / `rows` / `cols`, which would give us a bit more leeway in the future to make them not just iterators but whole array-like things.  I don't really like `rows` and `cols` in that case, but `slices` is actually quite nice.
2. Behavior: Are we ok with just a simple generator? That is, only exposing iteration as their API? Again, I like this as a starting point; generators have a _very small_ API surface and we could change the type in the future.

And, thirdly, on the implementation side, I'd like for us to spend a few cycles trying to find a type-stable implementation of `eachslice`.

arnavs commented:
1. What we could do is write `rows(A) = collect(eachrow(A))` and `cols(A) = collect(eachcol(A))` for now. And then, at some future point, maybe deprecate `eachrow` and `eachcol`.

andyferris commented:
> And, thirdly, on the implementation side, I'd like for us to spend a few cycles trying to find a type-stable implementation of eachslice.

It might be worth pointing ot that there is already a type-stable implementation [here](https://github.com/JuliaData/SplitApplyCombine.jl/blob/master/src/splitdims.jl#L75-L142) with (I think) all the features we want already - but I did have to cheat and use `return_type` to make it type stable, which isn't ideal.

@bramtayl uses `typeof` [here](https://github.com/bramtayl/JuliennedArrays.jl/blob/master/src/JuliennedArrays.jl#L20) instead. Is the compiler smart enough to infer the type and make this a no-op? What happens in the case the array is empty? (We don't want the empty case to throw an error, and neither do we want the empty case to impact type stability...)

In any case, I was hoping this would be more of a copy-paste job - with the challenging bit being the naming :laughing:.

> Behavior: Are we ok with just a simple generator? That is, only exposing iteration as their API? Again, I like this as a starting point; generators have a very small API surface and we could change the type in the future.

Sure, if we want a simple MVP then `Generator` seems OK? I don't think replacing this with an `AbstractArray` later should be very breaking at all.

bramtayl commented:
When I wrote the code I think I remember it being a no-op, which meant that that first just "disappeared". But looking back I wonder if that's contingent on the arrays having a fixed eltype?

bramtayl commented:
Also looking back I notice I should have added iteratorsize and iteratoreltype methods. it was never meant to be collected but I suppose someone might want to do that.

bramtayl commented:
I’m thinking now it should be fine pending a counterexample

datnamer commented:
We currently have to rewrite kernels for GPUs I think. Is there a way to make these iterators aware of the computational structure and automatically paralleize over distributed/shared /GPU arrays?

@SimonDanisch

Commit:
refactor eachslice; type-stable now?

arnavs commented:
@mbauman Implemented your suggestion (`eachslice(A, dim) = (selectdim(A, dim, i) for i in axes(A, dim))`), and I think this is type-stable?

andyferris commented:
I'm not great for `StaticArray`s, where `dim` needs to be a compile-time constant in the closure of the comprehension.

(Don't let this block the PR but I may have to work around this somehow).

AzamatB commented:
Can we merge this?

arnavs commented:
I asked @mbauman to take over on making the generic piece type-stable, since I think it's above my pay grade. 

If that looks like it'll take a while, maybe there's value in just merging the `AbstractVecOrMat` case for now.

mbauman commented:
I've not had a chance to do this yet — anyone is certainly welcome to take it over.  The key thing to check for type-stability is:

```julia
f(A) = first(eachslice(A, 1))
@code_warntype f(rand(3,3))
```

peterahrens commented:
I can't get this function to type-infer without an @inline annotation

```
"""
    eachslice(A::AbstractArray, dims)
Get an iterator over views of `A` along dimensions `dims`. If
`A` has less than `d` dimensions, `collect(eachslice(A, d))`
will just return a trivial collection with a view into `A`.
See also [`eachrow`](@ref) and [`eachcol`](@ref).
"""
@inline eachslice(A, dims) = _eachslice(A, Val(ndims(A)), Val(dims))

using Base.Iterators

@inline _splat_view(args) = view(args...)
@generated function _eachslice(A, ::Val{N}, ::Val{dims}) where {N, dims}
    slicer = [n in dims ? :(axes(A, $n)) : :((Colon(),)) for n in 1:N]
    return quote
        Base.@_inline_meta
        Base.Generator(_splat_view, Base.Iterators.product((A,), $(slicer...)))
    end
end

using InteractiveUtils

f(A) = first(eachslice(A, 1))
@code_warntype f(rand(3,3))
```

peterahrens commented:
@andyferris The current state of (lack of) specialization on keyword argument types, coupled with the design of functions like `mapslices(f, A, dims=dims)` and `cat(A..., dims=dims)`, represents a barrier to writing generated functions that can use constant-propagated values of `dims`. I recently had to implement a `mapslices` and a `cat` for `StaticArrays`, but I couldn't get the functions to be type-stable without deviating from the canonical `AbstractArray` definitions. Perhaps this is the wrong thread to discuss, but I think that an (optional) interface to these functions with `dims` as a positional argument could be really helpful. For instance, as we design the `eachslices` function signature, we must consider the same tradeoff. If `dims` is a keyword argument, we cannot make the function type-stable.

timholy commented:
This passes inference:
```julia
julia> @inline function eachslice(A, dim)
           dim <= ndims(A) || throw(DimensionMismatch("bad bad"))
           idx1, idx2 = ntuple(d->(:), dim-1), ntuple(d->(:), ndims(A)-dim)
           return (view(A, idx1..., i, idx2...) for i in axes(A, dim))
       end
```

timholy reviewed (commented):
I approve of the names. With tweaks and https://github.com/JuliaLang/julia/pull/29749#issuecomment-442571977 I think this could be merged.

arnavs commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,30 @@ function normalize(v::AbstractVector, p::Real = 2)
+eachcol(A::AbstractVecOrMat) = (view(A, :, i) for i in axes(A, 2))
+
+"""
+    eachslice(A::AbstractArray, d)
```
Good point.

arnavs commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,30 @@ function normalize(v::AbstractVector, p::Real = 2)
+"""
+    eachrow(A::AbstractVecOrMat)
+
+Get a generator over views of A's first dimension.
```
Agree.

arnavs replied:
Important to point out that they're views, though?

Commit:
git feedback

arnavs commented:
Thanks @timholy, incorporated your feedback.

fredrikekre commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,32 @@ function normalize(v::AbstractVector, p::Real = 2)
+"""
+    eachslice(A; dims = 1)
+
+Create a generator iterating over views of A's nth dimension.
```
Whats `n`?

arnavs replied:
Well, you can't really say `over A's dims-th dimension`...

AzamatB replied:
How about `over dimension dim of A`?

KristofferC replied:
It is good to refer to the argument in the function declaration.

fredrikekre commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,32 @@ function normalize(v::AbstractVector, p::Real = 2)
+Create a generator iterating over views of A's nth dimension.
+See also [`eachrow`](@ref) and [`eachcol`](@ref).
+"""
+@inline function eachslice(A; dims = 1)
```
Maybe `dims::Integer`?

arnavs replied:
Not sure I see the value. A non-integer `dims` would fail anyway.

mbauman replied:
It's also possible to eventually extend this implementation to iterating over many dimensions.  For example, `eachslice(reshape(1:24, 4, 3, 2); dims = (2, 3))` could meaningfully generalize to iterate over both the second and third dimension and thus return the columns.  Let's not worry about that for now — but let's give ourselves the space to develop that new capability by ensuring that it's an error to pass more than one dimension.  And let's support passing a tuple of one dimension — the keyword name is dim**s**, after all:

```julia
length(dims) == 1 || throw(ArgumentError("only single dimensions are supported"))
dim = first(dims)
# … remainder uses `dim` not `dims`
```

arnavs replied:
Done.

fredrikekre reviewed (dismissed):
I don't think these belong in `LinearAlgebra`.

arnavs commented:
@fredrikekre To be clear, where else would they belong?

KristofferC commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1415,3 +1415,32 @@ function normalize(v::AbstractVector, p::Real = 2)
+Create a generator iterating over views of A's nth dimension.
+See also [`eachrow`](@ref) and [`eachcol`](@ref).
+"""
+@inline function eachslice(A; dims = 1)
```
Since this is an untyped `A`, what is the interface that it needs to follow for it to be valid to call `eachslice` on it?

mbauman replied:
Good point: it's always easier to artificially restrict new functionality and later widen it.  Let's just assert `::AbstractArray` here.

arnavs replied:
Done.

mbauman commented:
Let's put it in abstractarraymath.jl right next to the `selectdim` function.

I always struggle with language around slices of dimensions vs. slices over dimensions.  E.g., what's a slice of `A`'s first dimension?  Is that a row or a column?  How about a slice over the first dimension?  Or a slice across the first dimension?  Getting a good, clear, concise docstring here is tricky.

How about:

* `eachrow(A)`
    >Creates a generator that iterates over the first dimension of matrix `A`, returning the rows as views.
* `eachcol(A)`
    > Creates a generator that iterates over the second dimension of matrix `A`, returning the columns as views.
* `eachslice(A; dims)`
    > Creates a generator that iterates over the given dimension of `A`, returning views that select all the data from the other dimensions in `A`.
    >
    > Equivalent to `(view(A,:,:,...,i,:,:,...)) for i in axes(A, dims))`, where `i` is in position `dims`.
    >
    > See also ``[`selectdim`](@ref)``

arnavs commented:
Sure, documentation looks great. Will make these changes.

fredrikekre commented:
> To be clear, where else would they belong?

`Base` somewhere (Matts proposal looks reasonable), it has nothing to do with LinearAlgebra at least.

Commit:
more git feedback

arnavs commented:
Responded to feedback above.

mbauman commented:
Awesome!  Looking really good.  One last thing on my account: tests. :)

Commit:
remove default for eachslice

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -406,3 +406,36 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
+eachcol(A::AbstractVecOrMat) = (view(A, :, i) for i in axes(A, 2))
+
+"""
+    eachslice(A; dims = 1)
```
I often try to place them near related functionality or in a test file that matches the source file.  Around here looks good:

```
$ ack selectdim test
test/arrayops.jl
1922:@testset "selectdim" begin
```

arnavs replied:
OK, put the tests right after.

Commit:
add tests

mbauman commented:
Crud, one problem:

```
julia> @eval Base begin
       eachrow(A::AbstractVecOrMat) = (view(A, i, :) for i in axes(A, 1))
       export eachrow
       end

julia> using DataFrames, RDatasets

julia> df = dataset("datasets", "iris");

julia> eachrow(df)
WARNING: both DataFrames and Base export "eachrow"; uses of it in module RDatasets must be qualified
WARNING: both DataFrames and Base export "eachrow"; uses of it in module Main must be qualified
ERROR: UndefVarError: eachrow not defined
Stacktrace:
 [1] top-level scope at none:0
```

arnavs commented:
Not too clear on the development process, but if this feature was included in 1.1 or 1.2, wouldn't that be enough?

mbauman commented:
Well, we're trying very hard to limit breaking changes in the 1.x releases. While we've said that exporting new names is an "ok" kind of breakage to have happen, this is a heavily-used core package and I don't want to make this change unless we have a plan in place for support/mitigation.

arnavs commented:
cc @ararslan 

I think there are two things here: 

1. For new versions of `DataFrames.jl`, we should do this properly (i.e., `import Base: eachrow, eachcol` or whatever)

2. For old versions of `DataFrames.jl`, can we retroactively change their compatibility to require Julia up to but not including 1.1?

andyferris commented:
> DataFrames can and should really extend this base name once it exists

I'm not sure if that is a bit of a pun? They could just use `Tables.rows`.

(In the context of IndexedTables and TypedTables, where a table can be an `AbstractArray`, they'll *definitely* need to respect the array meaning of `eachcolumn` for example, not the tabular meaning).

andyferris commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1414,4 +1414,4 @@ function normalize(v::AbstractVector, p::Real = 2)
         return T[]
     end
-end
+end
```
Did you mean to delete this terminating newline?

arnavs replied:
Honestly not too plugged into the whitespace changes. Will revert.

nalimilan commented:
> Edit: how about we pre-empt the problem with an advance PR to DataFrames that checks if `eachrow` and `eachcol` exist in Base and then imports them if so?

Yes, that sounds like the right solution. That's essentially equivalent to defining the functions in Compat and overloading them from DataFrames. Maybe the latter is preferable as it can be used by other packages. Anyhow I'm not worried about breakage, it doesn't really matter that old DataFrames releases don't fully work on Julia 1.1 (adding a new function is not a breach in the API stability contract).

arnavs commented:
OK @nalimilan or @mbauman, do you think you could submit such a PR? I'd do it, but there would probably be things to iterate on.

Commit:
prune eachrow eachcol exports and newline fix?

Commit:
typo in exports

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -406,3 +406,36 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
+eachcol(A::AbstractVecOrMat) = (view(A, :, i) for i in axes(A, 2))
+
+"""
+    eachslice(A; dims)
```
```suggestion
    eachslice(A::AbstractArray; dims)
```

arnavs replied:
True.

Commit:
update docs for AbstractArray [ci skip]

mbauman commented on test/arrayops.jl:
```diff
@@ -1932,6 +1932,21 @@ end
     @test IndexStyle(selectdim(A, 3, 1)) == IndexStyle(view(A, :, :, 1)) == IndexLinear()
 end
 
+# row/column/slice iterator tests
```
```suggestion
# row/column/slice iterator tests
using Base: eachrow, eachcol
```

mbauman committed:
Update test/arrayops.jl

Co-Authored-By: arnavs <soodarnav01@gmail.com>

nalimilan commented on base/abstractarraymath.jl:
```diff
@@ -406,3 +406,36 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
+"""
+    eachslice(A::AbstractArray; dims)
+
+Creates a generator that iterates over the given dims of A, returning views that select all the data from the other dimensions in A.
```
Break lines at 92 chars.

fredrikekre replied:
Also `Creates` -> `Create` etc.

arnavs replied:
Done.

nalimilan commented on base/abstractarraymath.jl:
```diff
@@ -406,3 +406,36 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
+"""
+    eachrow(A::AbstractVecOrMat)
+
+Creates a generator that iterates over the first dimension of matrix A, returning the rows as views.
```
```suggestion
Creates a generator that iterates over the first dimension of vector or matrix `A`,
returning the rows as views.
```
(same below, in particular for missing backticks)

arnavs replied:
Missed what you meant about the backticks, but I don't think this is worth running CI again over.

nalimilan replied:
I mean that there are more occurrences of code or object names that need backticks below. You can put `[ci skip]` in the commit message to avoid running CI on it.

nalimilan commented:
Don't worry about fixing DataFrames, we can do that on our end. But why did you remove `eachrow` and `eachcol` from exports?

fredrikekre commented:
> But why did you remove eachrow and eachcol from exports?

I proposed on slack (and I think Matt agreed?) that we leave these unexported for Julia 1.1 but document that they will be exported in Julia 1.2. Exporting new names is breaking and I think we should consider doing the same for other new exports. In this case DataFrames just happens to be a package that we know of, and can fix, but I think it would be nice to give one minor release cycle for new names as a heads up and then we can export them in the next release.

We can export them from the `Future` standard library so they are available after `using Future` in Julia 1.1.

nalimilan commented:
That sounds like a bit radical to me. In this case the conflict is obvious (since as you say DataFrames is a highly visible package), but would we be willing to apply this rule to any new export? Looking at NEWS.md, I see 1.1 will also introduce `splitpath` and `isnothing`.

There will be some time anyway between the feature freeze and the final release, so that leaves time for packages to spot that their tests are failing on nightlies and fix that. We could also have a policy to file issues against all packages which PkgEval detects as broken due to this, for example right after the feature freeze.

fredrikekre commented:
> but would we be willing to apply this rule to any new export? Looking at NEWS.md, I see 1.1 will also introduce splitpath and isnothing.

Yes, that is what I meant. We haven't really discussed the problem with new exports AFAIK and if we havr lived without these functions for years I think we can live 4 more months without them?

> We could also have a policy to file issues against all packages which PkgEval detects as broken

All Julia code out there is not PkgEvalable.

nalimilan committed:
Update base/abstractarraymath.jl

Co-Authored-By: arnavs <soodarnav01@gmail.com>

Commit:
bikeshedding

mbauman commented:
My thoughts are:
* we’re very close to 1.1
* fixing this _after_ we break it will lead to gripes and difficulties validating with PkgEval due to the many downstream dependents
* exporting from Future is cute, but it doesn’t really seem to gain us much over having folks know that they’re public in `Base` and will eventually be exported. 
* waiting until 1.2 to export them will give folks time to upgrade DataFrames.

nalimilan commented:
Honestly I don't think adding unexported functions in one release will help people to fix their code in advance. Only people following Julia development closely will be aware of that, and these people are precisely the ones which will adapt before 1.1 is out anyway (like DataFrames). Others will only notice the breakage in 1.2 when we eventually export these symbols.

Regarding DataFrames, let's not take it as an argument to delay the addition of new exports. @bkamins has already made a PR to make the API more consistent with this PR, and we can merge a fix even before this PR is merged in Julia, and tag it in a few days.

arnavs commented:
If everyone agrees that ‘eachslice’ is good to export, perhaps we could do that now and continue the discussion on the other two?

Not to put too fine a point on it, but I think we’ve reached the convergence point from this PR (the functionality we want is added, documented, tested, and won’t break things downstream). If I’m wrong, happy to iterate further, but this feature request has been open for years now and it would be nice to close.

mbauman commented:
I would just very much like to find a conclusion here. @Arnavs has been very tenacious, persistent and patient for this first PR. I leaned on the side of being conservative to help make that happen.

nalimilan committed:
Add missing backticks and improve message

[ci skip]

nalimilan commented:
AFAICT we all agree the PR can be merged. The debate is about whether we should add more exports or not. But if we don't reach an agreement soon better merge and continue the discussion after that.

bkamins commented:
If some package used e.g. `eachcol` earlier (like DataFrames.jl) the maintainers will have to have to introduce a conditional definition of `eachcol` and `Base.eachcol` anyway to keep supporting Julia 1.0 (or handle this duality via Compat.jl) no matter if we export it in Julia 1.1 or 1.2.

AzamatB commented:
My 2¢: I agree with @nalimilan and @bkamins and think there is no need to postpone exports to 1.2

Commit:
add exports for eachcol, eachrow

[ci skip]

arnavs commented:
Alright, I re-added the exports. Let's never think about this PR ever again :)

StefanKarpinski commented:
@nalimilan or @mbauman, please merge if and when you think this is ready (with appropriate squashing).

mbauman commented:
I'm thrilled DataFrames folks are on board here and will accommodate the breakage.  Thanks everyone. Let's merge!
